Nested `DropdownMenu` stays open at deeper levels

rohan-kadkol opened the pull request:
<!--
Thank you for your contribution to iTwinUI.

If you are only making changes to the docs site using the "Edit page on GitHub" link,
then you can ignore most of this template.
-->

## Changes

<!--
What kind of code changes does this PR include?
Mention anything that could be helpful for reviewers and include screenshots for visual changes.
-->

Fixes #1924 

WIP

## Testing

<!--
How did you test your changes?
If your PR has visual changes, then make sure they are demonstrated in css-workshop and react-workshop, then approve visual test images for both (`pnpm approve:css` and `pnpm approve:react`).

If not applicable, you can write "N/A".
-->

## Docs

<!--
If your PR includes user-facing changes, then update docs in all places (JSDoc, website, stories, etc).
Make sure to include a changeset (`pnpm changeset`).

If not applicable, you can write "N/A".
-->

mayank99 commented on packages/itwinui-react/src/core/Menu/MenuItem.tsx:
```diff
@@ -115,16 +127,54 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
+      isSubmenuVisible ||
+      isNestedSubmenuVisible ||
+      // to keep the submenu open when mouse enters it and then hovers out
+      dropdownMenuContext.lastHoveredNode?.parentId === nodeId,
```
I don't think this is quite what we discussed. By "hover ≈ focus" I meant literally focusing an item (by calling `.focus()`) when it is hovered. Arrow-key navigation should continue from the last hovered item.

mayank99 replied:
> But triggering :focus-visible and then hovering over Item 4_2 does not do that.

If by "triggering `:focus-visible`" you mean in dev-tools, then it's usually not representative of real interactions. Usually you want to check both `:focus` and `:focus-visible`. And even then it is mainly useful for debugging styles (CSS), not behavior (JS). For proper debugging, you can use the same snippet mentioned in https://github.com/iTwin/iTwinUI/issues/1924#issuecomment-2014097457.

Nothing stands out to me in code, but if I had to guess, it's likely the `hover`+`focus` triggers in `usePopover` that's messing with it.

mayank99 replied:
The `setFocusOnSubmenu` logic could also be conflicting with it. Ideally we wouldn't need to use this hack at all.

rohan-kadkol replied:
> By "hover ≈ focus" I meant literally focusing an item (by calling `.focus()`) when it is hovered

In [`4b29b10`](https://github.com/iTwin/iTwinUI/pull/1942/commits/4b29b1077625712e9997760558386df0b339f3d8), I removed the `useHover` trigger and but kept the `useFocus` trigger. Now, when we hover over a `MenuItem`, we call `.focus()` on itself inside the `onMouseEnter`. Since the `useFocus` trigger exists, hovering still indirectly opens a submenu.

Since "hover ≈ focus", one change is that earlier, opening a `DropdownMenu` where the first item has a submenu used to not open that submenu. But not since "hover ≈ focus", that submenu is opened. Not sure if this change in behavior is unideal.

We can see this when opening the leftmost menu button using a mouse click in the below recording:

<details>
<summary>Screen-recording</summary>

Video parts:
1. Second DropdownMenu mouse: no noticeable behavior change
2. First DropdownMenu mouse: `Item 1_3`'s submenu is not automatically open
3. Second DropdownMenu keyboard: no noticeable behavior change
4. First DropdownMenu keyboard: no noticeable behavior change

https://github.com/iTwin/iTwinUI/assets/45748283/aae38f59-907b-4f10-9946-f89209c03cf7

</details>

In the above-mentioned commit, I temporarily hard-coded [`visibleOnly`](https://floating-ui.com/docs/usefocus#visibleonly) to `false` in `Popover.tsx` to make this work. I will changing the hard coded value to be general if we decide to continue down this path.

> I had tried to call `.focus()` in `onMouseEnter` but kept getting unexpected behavior similar to this:

Like you suggested and I had alluded to before, the hover and focus triggers were for some reason not cooperating. The issue I mentioned is no longer present after removing the `onHover` trigger.

mayank99 replied:
The first submenu automatically opening when the main menu is opened is not acceptable. The user only meant to open the main menu. They should be expected to hover or arrow-right on the first item to open the submenu.

Ideally we'd be able to use [`focus({ focusVisible: true })`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/focus#focusvisible), but it has poor browser support.

So I think we'll need to look into the heuristics of how `visibleOnly` is implemented in floating-ui to figure out how we can emulate focusVisible.

rohan-kadkol replied:
I noticed that maintaining an `isHovered` state variable (https://github.com/iTwin/iTwinUI/pull/1942/commits/5ac32c6643f3f23c93b5c7b12b15b3f92ad0d953) solves the issue of unexpectedly closing all submenus upon hovering over "Item 4_2" as mentioned in https://github.com/iTwin/iTwinUI/pull/1942#discussion_r1556174966.

In the same commit, I also brought back the hover trigger. This helped bring back the [`safePolygon`](https://floating-ui.com/docs/usehover#safepolygon). Everything still continues to work.

> The first submenu automatically opening when the main menu is opened is not acceptable.

I also was able to undo setting `visisbleOnly: false` in `useHover` in `Popover.tsx`. So, "Item_3" in the leftmost DropdownMenu now no longer opens its submenu upon opening the DropdownMenu after clicking the "Menu" button. It only opens upon hovering "Item 1_3".

<details>
<summary>Screen-recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/3c7b68d0-f355-4232-b7dd-f531b16d048e

</details>

mayank99 replied:
> > The first submenu automatically opening when the main menu is opened is not acceptable.
> 
> I also was able to undo setting `visisbleOnly: false` in `useHover` in `Popover.tsx`. So, "Item_3" in the leftmost DropdownMenu now no longer opens its submenu upon opening the DropdownMenu after clicking the "Menu" button. It only opens upon hovering "Item 1_3".

I think the root cause may not be fixed. Because the same problem can happen inside a nested menu when the first item has another submenu.

<details><summary>Screen recording</summary>

In here, the third level menu automatically opens when Item #4 is focused.

https://github.com/iTwin/iTwinUI/assets/9084735/fa6491e2-3ca4-449f-8b5f-37740e09fda4

</details> 

To avoid these complications, I'm now rethinking whether we should keep `focus` as the trigger.

Arrow keys are already used as the trigger, and in the recent UX call we also discussed using click/<kbd>Enter</kbd> as the trigger (#879). Together, these two triggers might make the `focus` trigger not as important.

If this change is too complicated, it can also happen in a separate follow-up PR.

rohan-kadkol replied:
> In here, the third level menu automatically opens when `Item #4` is focused.

Ohh, I thought that was expected since we decided hover ≈ focus. As a result, since hovering over `Item #4` would open the third level submenu, focusing it also does the same.

> To avoid these complications, I'm now rethinking whether we should keep `focus` as the trigger.

Sounds good, this was my preference too because that is how the menus work in other UI libraries (e.g. [Radix UI](https://www.radix-ui.com/primitives/docs/components/dropdown-menu), [React Aria](https://react-spectrum.adobe.com/react-aria/Menu.html#submenus), [AriaKit](https://ariakit.org/examples/menu-nested)). Even the menu bar menus and context menus on my Mac work the same way (i.e. doesn't open the submenu on focus and only opens them using the <kbd>←</kbd>/<kbd>→</kbd> keys).

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/060fcda7-a378-495c-a205-5e7bb0c21d7a

</details>

I guess they might have been working differently (i.e. using focus as a trigger) on your Mac because of some change in your settings.

> If this change is too complicated, it can also happen in a separate follow-up PR.

Great, so, I will try working on removing the focus trigger and seeing how it works and also see whether a new PR would be better.

mayank99 replied:
> Ohh, I thought that was expected since we decided hover ≈ focus. As a result, since hovering over Item #4 would open the third level submenu, focusing it also does the same.

You can think of it as a one-directional relationship. Hovering should focus the item, but focusing by itself is its own separate thing that does not need to behave.

The more important thing to remember is the user's intent. Opening a menu automatically focuses the first menu-item, but this focus was not moved by the user. So they didn't intend to open the sub-menu. Compare this to a mouse user who needs to intentionally move their mouse from the trigger element into menu-item.

> Even the menu bar menus and context menus on my Mac work the same way … I guess they might have been working differently (i.e. using focus as a trigger) on your Mac because of some change in your settings.

I couldn't find any setting that I changed. It's very possible that the behavior was changed in macOS 14. I prefer it this way, but it very much needs to be done properly (which currently it is not).

> Great, so, I will try working on removing the focus trigger and seeing how it works and also see whether a new PR would be better.

If you intend to do a separate PR for the click trigger, then I would keep the focus trigger unchanged in this PR.

rohan-kadkol replied:
> You can think of it as a one-directional relationship. Hovering should focus the item, but focusing by itself is its own separate thing that does not need to behave identically. 

Sounds good, with the recent changes, focusing a MenuItem does not open its submenu until <kbd>→</kbd> is pressed:

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/f2e8b202-b2fb-4678-8648-3e80420d4bcf

</details>

In contrast, hovering a MenuItem opens its submenu:

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/c0e1c690-a20a-4951-b7ba-0e7559e69255

</details>

> If you intend to do a separate PR for the click trigger, then I would keep the focus trigger unchanged in this PR.

Yes, I hope to work on the click trigger in a separate PR since that required more work than just adding the trigger.

I have currently removed the focus trigger to stop opening the submenu on focus, as mentioned in the point above. Is there any strong reason to keep the focus trigger in this PR instead of removing it?

rohan-kadkol replied:
Adding back `focus` takes away some of the benefits from this PR. E.g. pressing <kbd>←</kbd> on a menu "X" will move focus to X's parent but still keep X open instead of closing it.

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/05dffd03-bedf-4299-8cf8-fc633eed2094

</details>

> When both focus and click are absent, arrow keys are the only trigger, which is not sufficient.

To confirm, why are arrow keys and hover insufficient triggers?

mayank99 replied:
> To confirm, why are arrow keys and hover insufficient triggers?

With complex patterns like submenus, there should always be multiple affordances. A screen-reader or keyboard user might not even realize that they need to press the arrow keys to open submenus.

In an ideal scenario, we would have focus+click+arrow-keys as triggers (+hover for mouse users). The next best thing is click+arrow-keys (+hover).

("click" also implies <kbd>Enter</kbd>/<kbd>Space</kbd> keypress)

rohan-kadkol replied:
Sounds good, added the click trigger manually instead of using `useClick` from floating UI (https://github.com/iTwin/iTwinUI/pull/1942/files/1b5f5f2f1f88375ebd8fe76b45887dc241d3ab54..a9d865f26b10f759df61210091fb0c0d00a16108).

So, the triggers now are hover, click (including <kbd>Enter</kbd>/<kbd>Space</kbd>), and arrow keys.

mayank99 commented on packages/itwinui-react/src/core/Menu/MenuItem.tsx:
```diff
@@ -159,92 +206,129 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
       }
       case 'Escape': {
-        // focus might get lost if submenu closes so move it back to parent
-        parent.ref?.current?.focus();
+        dropdownMenuContext.close();
+
+        event.stopPropagation();
+        event.preventDefault();
```
I think this whole branch might be unnecessary if we use [`bubbles`](https://floating-ui.com/docs/usedismiss#bubbles)

rohan-kadkol replied:
Ohh, nice, hardcoded `bubbles: true` in `Popover.tsx` to get the same behavior as this branch. (https://github.com/iTwin/iTwinUI/pull/1942/commits/e6b628bd4cae8a3bd855b1107e31032faf378434).

Then in [`271c4bc`](https://github.com/iTwin/iTwinUI/pull/1942/commits/271c4bc791722d5acddba972c39c36e0e6c46465), I added a new prop to `usePopover` to add `bubbles: true` only for `MenuItem` and not for other uses of `usePopover`.

mayank99 replied:
The prop is not needed. Menu is the only place we use `FloatingTree`, so there are no "other uses". And even if we add `FloatingTree` in more places, I'd argue we want the bubbles behavior in all of those places too.

The one exception is tooltips. I don't think that dismissing a tooltip using <kbd>esc</kbd> should close other popovers. But since Tooltip has its own `useFloating` call, I believe this should already work fine.

rohan-kadkol replied:
One more case I felt where `bubbles: true` was not working well was in `DatePickerInput`. Because <kbd>esc</kbd> was closing the entire filter instead of just the `DatePicker`.

<details>
<summary>bubbles: true/false</summary>

`bubbles: true`

https://github.com/iTwin/iTwinUI/assets/45748283/3041105d-0983-4f1d-967f-fbefbf01b9ba

`bubbles: false`

https://github.com/iTwin/iTwinUI/assets/45748283/4f3dedf6-1a6b-448b-9513-afc6f3820b7e

</details>

> The one exception is tooltips. I don't think that dismissing a tooltip using esc should close other popovers. But since Tooltip has its own `useFloating` call, I believe this should already work fine.

Yes, I just confirmed that `bubbles: true` does not affect the behavior when a tooltip (e.g. from an `IconButton`) is in a `Dialog`.

<details>
<summary>Screen-recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/bd171f76-4967-4e42-aca0-48cc14c89151

</details>

mayank99 replied:
Oh interesting, in that case I think it should be dependent on whether there is a FloatingTree.

If usePopover called in a component within a FloatingTree, then it should bubble. If not, then don't bubble.

rohan-kadkol replied:
Ohh, sounds good. Removed `triggerProps` and now only bubbling when inside a `<FloatingTree>` scope. (https://github.com/iTwin/iTwinUI/pull/1942/commits/5dac2d3aacf9ebec50c301cfa18e72857a30937d)

AdamMeza-Bentley commented on packages/itwinui-react/src/core/Menu/MenuItem.test.tsx:
```diff
@@ -265,20 +265,19 @@ it('should handle key press with sub menus', async () => {
   expect(screen.queryByTestId('sub')).toBeFalsy();
 
-  // go right to open sub menu
+  // going right to should re-open the sub menu and move the focus to the first item in the submenu.
```
Extra 'to'

```suggestion
  // going right should re-open the sub menu and move the focus to the first item in the submenu.
```

AdamMeza-Bentley commented on packages/itwinui-react/src/core/Menu/MenuItem.test.tsx:
```diff
@@ -288,3 +287,52 @@ it('should handle key press with sub menus', async () => {
+  // focus to open sub menu
+  act(() => menuItem.focus());
+
+  // go right to move focus to the first submenu item. The first submenu's submenu should open.
```
Similar to the above comment. I don't know if the wording is intentional but it feels a bit clunky.

rohan-kadkol replied:
We could try combining the two sentences:

```suggestion
  // go right to move focus to the first submenu item and open its submenu.
```

… but I'm not sure if it makes it clear that the submenu's submenu should open. i.e. what are we testing for.

Or we could split it on two lines.

```diff
- // go right to move focus to the first submenu item. The first submenu's submenu should open.
+ // Go right to move focus to the first submenu item.
+ // The first submenu's submenu should open.
await userEvent.keyboard('{ArrowRight}');
expect(screen.getByTestId('sub-1')).toHaveFocus();
expect(screen.getByTestId('sub-2')).toBeVisible();

- // hover over the second submenu item. The first submenu's submenu should close.
+ // Hover over the second submenu item.
+ // The first submenu's submenu should close.
fireEvent.mouseEnter(screen.getByTestId('sub-3'));
expect(screen.queryByTestId('sub-4')).toBeFalsy();
```

Or maybe do you have any other suggestions? Because I'm not quite sure I understood what you meant by "clunky".

AdamMeza-Bentley replied:
I think the first option is good! I guess more specially the grammar seemed a bit off. But the first option is great

rohan-kadkol replied:
Great, combined the two sentences similar to the first option (https://github.com/iTwin/iTwinUI/pull/1942/commits/c507a95f33585f9c25a646d8b4f0ff42c82e9fbb).

rohan-kadkol commented on packages/itwinui-react/src/core/Menu/MenuItem.tsx:
```diff
@@ -159,92 +206,123 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
+ *
+ * This is useful when `MenuItem`s are used outside of a `DropdownMenu`.
+ */
+const ConditionalFloatingTreeWrapper = (props: {
```
I noticed that `MenuItem`s were used outside of `DropdownMenu` in the unit tests in `MenuItem.test.tsx`. So, I thought to have something like `ConditionalFloatingTreeWrapper` to add a `FloatingTree` when it does not exist.

Open to other ideas to implement this.

mayank99 replied:
> `MenuItem`s were used outside of `DropdownMenu` in the unit tests in `MenuItem.test.tsx`.

Why is this a problem? It should be fine.

Also we should almost never need to change our code for the purpose of making unit tests pass.

rohan-kadkol replied:
> Why is this a problem? It should be fine.

We are currently relying on `FloatingTree` + `FloatingNode` to send events when a submenu opens. These events help us enforce only one submenu open at a time for each menu. Since tree events are not possible without a `FloatingTree` context as an ancestor, I added `ConditionalFloatingTreeWrapper`.

> Also we should almost never need to change our code for the purpose of making unit tests pass.

True, but I was trying to catch even non-unit test cases where `MenuItem`s could be outside the `DropdownMenu`.

If this case doesn't exist, then I can remove `ConditionalFloatingTreeWrapper` and update the unit tests accordingly.

mayank99 replied:
It might be better to test these interactions in a playwright test, with the whole component. Unit tests are a bad place to test such important, complex logic.

I don't think `MenuItem` should be used anywhere else. The only place I can think of is in the custom item renderers of `Select` and `ComboBox`, but those should use `ListItem` imo. We might want to look into showing a runtime warning to enforce this.

rohan-kadkol replied:
> I don't think `MenuItem` should be used anywhere else.

For now, I removed `ConditionalFloatingTreeWrapper` since one more disadvantage of using it is that the top-level menu will not have one FloatingTree. Instead, each top-level menu's MenuItems will have a FloatingTree each. So, enforcing just one active submenu for the first level menu will not be possible with this approach.

> Or even deprecate `MenuItem` altogether and replace it with something like `DropdownMenu.Item`.

If we deprecate `MenuItem`, do we still need to support it until it is removed? If so, then the problem will exist for users that are using `MenuItem` instead of `DropdownMenu.Item`.

> The only place I can think of is in the custom item renderers of `Select` and `ComboBox`, but those should use `ListItem` imo. We might want to look into showing a runtime warning to enforce this. 

`Select` and `ComboBox` would not have submenu items. And the floating tree events are only for submenu management. So, even if the `<FloatingTree>` wrapper is not present, I don't think it would be a problem.

To be safer, should/can we add a `<FloatingTree>` inside `Select` and `ComboBox`, similar to how we're doing for `DropdownMenu`?

mayank99 replied:
> To be safer, should/can we add a `<FloatingTree>` inside `Select` and `ComboBox`, similar to how we're doing for `DropdownMenu`?

No, there's no need. I don't see how that's safer, since `ComboBox`/`Select` don't have submenus.

> If we deprecate `MenuItem`, do we still need to support it until it is removed? If so, then the problem will exist for users that are using `MenuItem` instead of `DropdownMenu.Item`.

Yes we need to support it, but that's not a difficult thing (I'm not sure what you mean by "the problem").

The point of a deprecation here is mainly to enforce correct usage. Currently if you use `MenuItem` inside `ComboBox`/`Select`, it gets incorrect semantics and needs to execute a bunch of unrelated code that gets thrown away anyway (because it is only relevant for `DropdownMenu`).

rohan-kadkol replied:
> I'm not sure what you mean by "the problem"

By "the problem", I meant that even if we deprecate `MenuItem` inside`Select`/`ComboBox`, if a user still uses a `MenuItem` there, enforcing only one active submenu per menu will not be possible without wrapping the entire tree in `FloatingTree`. But I guess that since submenus should anyways not be there in `Select`s/`ComboBox`es, "this program" will not happen.

rohan-kadkol commented on packages/itwinui-react/src/core/Menu/MenuItem.tsx:
```diff
@@ -107,27 +106,119 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
+
+  const [rightArrowPressed, setRightArrowPressed] = React.useState(false);
+
+  // Needed to trigger arrowRightPressed only after the FloatingTree has been
+  // updated with the nodes from the new submenu.
+  React.useEffect(
+    () => {
+      if (
+        rightArrowPressed &&
+        subMenuItems.length > 0 &&
+        !!tree?.nodesRef.current.find((n) => n.parentId === nodeId) // Tree has been updated with the new submenu
+      ) {
+        tree.events.emit('arrowRightPressed', {
+          nodeId,
+          parentId,
+        } satisfies TreeEvent);
+
+        setRightArrowPressed(false);
+      }
+    },
+    // TODO: Try to not require this eslint disable
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [
+      rightArrowPressed, // Run this whenever rightArrowPressed changes
+      nodeId,
+      parentId,
+      subMenuItems.length,
+      tree?.events,
+      // TODO: How to handle this warning?
+      // warning  React Hook React.useEffect has an unnecessary dependency: 'tree?.nodesRef.current'. Either exclude it or remove the dependency array. Mutable values like 'tree.nodesRef.current' aren't valid dependencies because mutating them doesn't re-render the component  react-hooks/exhaustive-deps
+      tree?.nodesRef.current,
+      children,
+      tree?.nodesRef,
+    ],
+  );
```
I removed the `focusOnSubmenu` variable used to focus the first item in the submenu when <kbd>→</kbd> was pressed. This was because it was causing focus issues after the changes from this PR. (Related: https://github.com/iTwin/iTwinUI/pull/1942#discussion_r1557993045)

As a replacement, I used this `useEffect`. This `useEffect` emits the `arrowRightPressed` event after certain conditions are met instead of directly emitting the `arrowRightPressed` event right after <kbd>→</kbd> is pressed. This is because we need to wait for the `FloatingTree`'s context to be updated with the new `FloatingNode`s from the submenu since `handleArrowRightPressed` needs the up-to-date tree to properly focus the first item.

This `useEffect` feels a bit hacky. Also, I disabled eslint for the deps array; not sure if that's alright. So, open to suggestions/alternatives.

mayank99 commented on packages/itwinui-react/src/core/Menu/MenuItem.tsx:
I'm a little confused by what happened here. Last time I looked at this PR (last week), I think it was in a pretty ok state with some small changes pending. But now, there's suddenly 100+ new lines of (hacky) code that don't feel right. What's the main difficulty here that is necessitating such code? Are we picking the wrong abstractions? Do we instead need to rewrite the DropdownMenu/Menu/MenuItem components in a way that makes it less difficult to achieve the behavior we desire?

rohan-kadkol replied:
> What's the main difficulty here that is necessitating such code?

When testing with edge cases, there were a few issues coming up. And it was difficult to fix them without replacing some of the previous workaround/hacky code (e.g. `focusOnSubmenu`) with something else. That's why I had to change my approach in order to make it work in the edge cases.

> Are we picking the wrong abstractions? Do we instead need to rewrite the DropdownMenu/Menu/MenuItem components in a way that makes it less difficult to achieve the behavior we desire?

I don't think rewriting it any more is necessary, at least at the moment since everything seems to be working well at the moment. So, for now, some alternatives to the hacky parts would be sufficient.

However, when working on the Layered Dropdown Menu from #653, a re-write might be needed; but not at the moment.

mayank99 replied:
The rewrite _is_ "alternative to the hacky parts". We can't keep adding more hacky code on top of the existing hacky code\*; that's how we end up with an unmaintainable codebase.

\* I've made it clear multiple times that the way submenus have been implemented (before this PR) is super hacky and might require a rewrite to truly fix.

---

At the very least, could you try to abstract away the hairy parts into private hooks/components to make the main components easier to understand?

rohan-kadkol replied:
> The rewrite _is_ "alternative to the hacky parts". We can't keep adding more hacky code on top of the existing hacky code*; that's how we end up with an unmaintainable codebase.
> 
> * I've made it clear multiple times that the way submenus have been implemented (before this PR) is super hacky and might require a rewrite to truly fix.

I guess I'm not sure about the scope of the rewrite we were hoping for. Because I did remove the hacky parts from before with improvements, of which only the useEffect that I mentioned in https://github.com/iTwin/iTwinUI/pull/1942#discussion_r1565664450 seems hacky.

mayank99 replied:
Almost the entirety of `Menu`/`MenuItem` was and continues to be hacky.

In addition to the multiple `useEffect`s, the other parts that can be improved include:
- The code structure and the split of responsibilities between `Menu`/`DropdownMenu`/`MenuItem` (each relying on implementation details of the others without being properly abstracted)
- The `isSubmenuVisible || <something-else>` condition (kinda ironic that `isSubmenuVisible` isn't the source of truth)
- The `rightArrowPressed` state variable (this is probably the biggest issue; it doesn't make sense to store this in state)
- The whole `isAncestor` function being executed on every render (performance issue, but also possibly indicative of a missing intermediate abstraction)

rohan-kadkol replied:
In the last few commits, I tried to make it less hacky. I used ideas from [floating-ui/floating-ui#1943](https://redirect.github.com/floating-ui/floating-ui/discussions/1943) to help with it.

There are still some loose ends, but here are some of the code improvements so far while still mostly maintaining the previous behavior:
* `isSubmenuVisible` is now the source of truth.
* `rightArrowPressed` and the associated `useEffect` are removed.
* `isAncestor` is removed.

rohan-kadkol commented on testing/e2e/app/routes/DropdownMenu/spec.ts:
```diff
@@ -0,0 +1,62 @@
+import { test, expect } from '@playwright/test';
+
+test.describe('DropdownMenu', () => {
+  test('should support deep level submenus', async ({ page }) => {
```
This e2e test seems to be flaky. Any suggestions on how to make it pass reliably?

mayank99 replied:
Have you tried debugging it using [`--ui` mode](https://playwright.dev/docs/test-ui-mode)?

Full command:
```
pnpm run build --filter=e2e && pnpm --filter=e2e run test --ui
```

After that, you could play with some delays. Playwright has multiples ways of introduces delays in various places ([example](https://playwright.dev/docs/api/class-keyboard#keyboard-press-option-delay)).

rohan-kadkol replied:
Ohh, I see, thanks. The UI mode helped me debug the issue and delays were one of the things that fixed the flakiness (https://github.com/iTwin/iTwinUI/pull/1942/files/7a41f31b492eb3ab656ecb1f8b43bb8bdbc2ac32..9e471ef813e96edd4ef841d7fae70f1c18b6e999).

mayank99 commented on packages/itwinui-react/src/core/Menu/MenuItem.tsx:
```diff
@@ -187,7 +190,7 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
       case ' ':
       case 'Spacebar': {
-        !disabled && onClick?.(value);
+        onActivate();
```
Can you remind me what was the problem with `useClick`?

_(Continued from [previous conversation](https://github.com/iTwin/iTwinUI/pull/1942#discussion_r1572609858))_

rohan-kadkol replied:
After adding the click trigger, I noticed that it required two clicks to close a submenu. But turns out that it was expected behavior ([floating-ui/floating-ui#1893 (comment)](https://redirect.github.com/floating-ui/floating-ui/issues/1893#issuecomment-1250161220)).

I worked around this by manually toggling the visible state in `onClick()` ([code](https://github.com/iTwin/iTwinUI/pull/1942/commits/b3c54a164a3f9e4c2760a4623031d6ae63f9e4de#diff-4410321a0ef9f99d8d2c38b7907baf2e10f7fdf8efcff3a9ca5245db96d35db9R223)).

mayank99 commented on packages/itwinui-react/src/core/Menu/MenuItem.tsx:
```diff
@@ -217,8 +220,20 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
+  const onClick = (e: React.MouseEvent<HTMLElement>) => {
+    onActivate();
+
+    e.preventDefault();
+    e.stopPropagation();
```
It would be useful to have a code comment for every instance of `preventDefault()` and `stopPropagation()`, because both of these are potentially destructive actions.

rohan-kadkol replied:
Upon testing, I realized that `e.preventDefault()` and `e.stopPropagation()` were not needed in this case. So, I removed them (https://github.com/iTwin/iTwinUI/pull/1942/commits/5ed937835d330eef33e3294e8d3ce9fe2f416506).

mayank99 commented on packages/itwinui-react/src/core/Popover/Popover.tsx:
```diff
@@ -98,32 +107,76 @@ type PopoverInternalProps = {
+} & PopoverInteractionProps &
+  Record<string, any>;
+
+/** If listNavigation is an interaction, it forces the required props for it to be provided */
+type PopoverInteractionConditionalProps =
+  | {
+      interactions?: {
+        listNavigation: true;
+      };
+      interactionsProps: {
+        listNavigation: UseListNavigationProps;
+      };
+    }
+  | {
+      interactions?: {
+        listNavigation?: false;
+      };
+      interactionsProps?: {
+        listNavigation?: UseListNavigationProps;
+      };
+    };
+
+type PopoverInteractionProps = PopoverInteractionConditionalProps & {
+  interactionsProps?: {
+    hover?: UseHoverProps<ReferenceType>;
+  };
 };
```
You shouldn't need all these complex types. It would be enough for every interaction to accept an object in addition to a boolean

Usage:
```js
interactions: {
  hover: { move: false },
  click: true,
}
```

mayank99 replied:
Why is it so complicated with the merge helper?

Something like this should be enough, I think (untested):

```js
useClick(
  floating.context,
  { enabled: !!interactions.click, ...interactions.click }
),
// …
```

The `!!` will coerce the objects into a boolean; and the `...` will automatically merge the objects (correctly overriding the defaults, while also being a noop if a boolean is used).

rohan-kadkol replied:
Since `interactions.click` can also be a `boolean` it gives a type error.

<img width="469" alt="image" src="https://github.com/iTwin/iTwinUI/assets/45748283/54127d15-161a-47c3-945f-7cb76ed9098b">

I can ignore the type error if we're okay with it because like you said, applying spread on a boolean will be no-op in JS.

mayank99 commented on packages/itwinui-react/src/core/Popover/Popover.tsx:
```diff
@@ -98,32 +107,76 @@ type PopoverInternalProps = {
    */
   matchWidth?: boolean;
+} & PopoverInteractionProps &
+  Record<string, any>;
```
nit: This `any` was meant to be a placeholder during our pair programming session. Can be replaced with the "correct" type now (if possible).

rohan-kadkol replied:
Replaced with proper type (https://github.com/iTwin/iTwinUI/pull/1942/commits/99f3b3a580c8aaffb556f85b6c331d32e72a7dad).

mayank99 replied:
I meant proper types for all arbitrary props (i.e. `UseFloatingProps`), rather than explicitly listing `nodeId`.

rohan-kadkol replied:
Even though `usePopover` is private, I still thought it would be good to keep the parameters list small. I thought that we can add more props as and when we need them.

But I can add all `useFloating`'s props (`UseFloatingProps`) now too (https://github.com/iTwin/iTwinUI/pull/1942/commits/82850b197e18f1b7386fc73c62705498a6284187).

mayank99 commented on packages/itwinui-react/src/core/Menu/MenuItem.tsx:
```diff
@@ -139,112 +190,143 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
+          subMenuItems.length > 0 ? popover.refs.setReference : null,
+          (el) => {
+            if (parent != null && parentTreeIndex != null) {
+              parent.listItemsRef.current[parentTreeIndex] = el as HTMLElement;
```
Can you try to explain in some more detail the new changes in this file since last week? I don't see them explained anywhere, and I'm very confused about how you're handling indexes.

Alternatively, I can reach out privately if you prefer to explain 1:1. After that we can try to improve the PR description and add some code comments.

mayank99 commented on packages/itwinui-react/src/core/Menu/MenuItem.tsx:
```diff
@@ -139,112 +190,143 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
+              >
+                {subMenuItems.map((item, index) =>
+                  React.cloneElement(item, {
+                    key: index,
```
You shouldn't need to set `key` here. `key` generally belongs to the user of the component.

rohan-kadkol replied:
Ohh, I see, removed (https://github.com/iTwin/iTwinUI/pull/1942/commits/e85db420d91fa8db90562d4d3ac398a2d5389496).

mayank99 commented on packages/itwinui-react/src/core/Menu/MenuItem.tsx:
```diff
@@ -139,112 +190,143 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
+                {subMenuItems.map((item, index) =>
+                  React.cloneElement(item, {
+                    key: index,
+                    ['data-menu-child-index']: index,
```
Two comments here:
1. Data attributes should be namespaced. Something like `data-iui-index` would suffice.
2. While it works for `MenuItem`, cloning elements like this is not super reliable. It would be better to use something like context. We could maybe save this improvement for when we deprecate `MenuItem` in favor of `DropdownMenu.Item` _([related conversation](https://github.com/iTwin/iTwinUI/pull/1942#discussion_r1561066189))_.

rohan-kadkol replied:
For now, I scoped the data attribute to `iui` by renaming it to `data-iui-index` ([`c3f0946`](https://github.com/iTwin/iTwinUI/pull/1942/commits/c3f09463b8bc7313883a36527d56506080cf2adc) + https://github.com/iTwin/iTwinUI/pull/1942/commits/8016f5bfcd8961c67ddb6a968094324157cd48e6).

I added an after PR TODO to try replacing the `cloneElement` with contexts.

mayank99 replied:
`cloneElement` and data attributes were replaced with context in [`142a531`](https://github.com/iTwin/iTwinUI/pull/1942/commits/142a531fa9b3dc358340aec0485405d217557419) in order to avoid reading from refs (and DOM elements) during render.

The other part of my comment (deprecating `MenuItem`) can be kept as an after-PR TODO.

mayank99 commented on packages/itwinui-react/src/core/Menu/MenuItem.tsx:
```diff
@@ -106,28 +116,69 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
+    },
+    interactionsProps: {
+      hover: {
+        move: false,
```
Would this be better suited as the default for `usePopover`? Do we ever need `move: true`?

rohan-kadkol replied:
I thought to avoid that to prevent causing any behavior changes for users already used to `move: true`. Also, I'm not sure in which cases would `move: true` be needed. But since Floating UI has set that as the default, I guess there is some reason for it?

I can add an after PR TODO to address this comment later or I can make the change in this PR itself if we think it won't impact users.

mayank99 replied:
Since it's a very small change, I'm ok with doing it in the same PR. Just make sure to check that there is no adverse behavioral effect, and add a short changeset as well.

rohan-kadkol replied:
Sounds good, I made that change in this PR (https://github.com/iTwin/iTwinUI/pull/1942/commits/93e5a33eed228c1cb43da0b50da9e4db062dc30d).

Since we don't expose the hover trigger in the public `Popover` and since we use the private `usePopover`'s `hover` trigger only in `MenuItem`, I think the existing changeset is sufficient ([code](https://github.com/iTwin/iTwinUI/pull/1942/files#diff-89ddc481dba81a1393f7122d869df11bb687a93faec74c18833d53f9399b297aR5)).

mayank99 commented on packages/itwinui-react/src/core/Popover/Popover.tsx:
```diff
@@ -154,15 +202,32 @@ export const usePopover = (options: PopoverOptions & PopoverInternalProps) => {
+      enabled: !!mergedInteractions.hover,
       delay: 100,
-      handleClose: safePolygon({ buffer: 1 }),
+      handleClose: safePolygon({ buffer: 1, requireIntent: false }),
```
While not directly relevant for this PR, I noticed safe polygon only works for the very last item in a menu. I investigated further and found that it starts working correctly when [`position: relative`](https://github.com/iTwin/iTwinUI/blob/e2e6ba7abbb46ac99ed225775b56942b88d8da04/packages/itwinui-css/src/menu/mixins.scss#L46) is removed. 


Further investigation needed. (I believe the `position: relative` is necessary for [`ListItem.Action`](https://itwinui.bentley.com/docs/list#with-listitemaction) to work as intended).

mayank99 commented on packages/itwinui-react/src/core/Menu/MenuItem.tsx:
```diff
@@ -139,112 +197,134 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
+    } satisfies TreeEvent);
+  };
+
+  const onClick = () => {
+    onActivate();
+  };
+
+  const onActivate = () => {
+    !disabled && subMenuItems.length === 0 && onClickProp?.(value);
+    setIsSubmenuVisible((prev) => !prev);
+  };
```
1. Since `onActivate` and now identical, they can be combined.
2. We should call `onClickProp` even when `subMenuItems` is provided, in case users were relying on that behavior.
3. We should not call `setIsSubmenuVisible` on disabled items because #1925`.
   - Additionally, we might also want to disable all other triggers on disabled items to make it consistent.

rohan-kadkol replied:
1. Combined (https://github.com/iTwin/iTwinUI/pull/1942/commits/a9ab28f140a4e8d0bb3d63e08735697a093f900e).
2. Ohh, sounds good, updated (https://github.com/iTwin/iTwinUI/pull/1942/commits/72058bb581bccdfc6867fc3ec4b0190faa5cd193).
3. I thought we planned to handle disabled `MenuItem`s later, but yes, reading your comment on the issue you linked, it feels better to handle it here. Handled (https://github.com/iTwin/iTwinUI/pull/1942/files/72058bb581bccdfc6867fc3ec4b0190faa5cd193..c556efa6d4ab39f823f80930084462a98303b7e7). These commits also disable all Floating UI triggers/interactions and our event handlers for disabled items.

mayank99 replied:
Why do you need to disable everything in so many places? We don't need all this extra code. It would be enough just _not_ render the submenu when the triggering menu-item is disabled.

rohan-kadkol replied:
> It would be enough just _not_ render the submenu when the triggering menu-item is disabled.

I did it this way since I thought we wanted to keep `isSubmenuVisible` as the source of truth (https://github.com/iTwin/iTwinUI/pull/1942#discussion_r1567940451). But I guess that discussion was only for non-disabled `MenuItem`s.

So, I replaced that code with the safer approach of not rendering the submenu when the triggering `MenuItem` is disabled (https://github.com/iTwin/iTwinUI/pull/1942/commits/6e39132e65d42d9f6c972c0d729ecb755a7067b3).

mayank99 commented on packages/itwinui-react/src/core/Menu/Menu.tsx:
```diff
@@ -89,6 +89,16 @@ export const Menu = React.forwardRef((props, ref) => {
+  // Keep focusedIndex in sync with the real focused node's index.
+  const onFocus = (event: React.FocusEvent<HTMLDivElement, Element>) => {
+    const items = getFocusableNodes();
+    const currentIndex = items.findIndex((el) => el === event.target);
```
Running `getFocusableNodes` and `findIndex` every time focus moves is expensive.

Suggestion to improve performance:
1. In `MenuItem`, set `data-iui-index={indexInParentTree}`.
2. In this function, read `e.target.dataset.index`.

rohan-kadkol replied:
I believe I had made this change before using [`useListNavigation`](https://floating-ui.com/docs/useListNavigation). Back then, I was instead relying on `Menu` for the <kbd>↑</kbd>/<kbd>↓</kbd> arrow keys navigation.

But since I'm now using `useListNavigation` for keyboard navigation, the code changes in `Menu.tsx` are no longer necessary. So, I reverted all changes in `Menu.tsx` (https://github.com/iTwin/iTwinUI/pull/1942/commits/ffb4fa269f78a91bb87a21047e2156009ba4971d).

rohan-kadkol commented on packages/itwinui-react/src/core/Menu/MenuItem.tsx:
```diff
@@ -139,112 +201,139 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
-      }
-      case 'Escape': {
-        // focus might get lost if submenu closes so move it back to parent
-        parent.ref?.current?.focus();
-        break;
-      }
-      default:
-        break;
     }
```
@mayank99 and I had removed this in a previous commit (https://github.com/iTwin/iTwinUI/pull/1942/commits/3861266a09f6d06d6a5f79e6e77261b56b12f309). But I brought it back since I now remember why I had it in the first place. It was because if we don't update the active index upon hover/focus, then Floating UI's `useListNavigation` keeps the stale `activeIndex`.

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/29fb5fc2-3ce0-4761-991f-6800f6edca9b

</details>

One small difference is that instead of updating the parent's activeIndex in `onFocus`, I am now doing it in `onMouseEnter`, where it is really needed because that is where we're manually changing the focus instead of letting `useListNavigation` handle it.

rohan-kadkol commented on packages/itwinui-react/src/core/Menu/MenuItem.tsx:
```diff
@@ -139,112 +201,139 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
-          event.stopPropagation();
-        }
-        break;
+      // Since we manually focus the MenuItem on hover, we need to manually update the active index for
+      // Floating UI's keyboard navigation to work correctly.
+      if (parent != null && indexInParentTree != null) {
+        parent.setActiveIndex(indexInParentTree);
       }
```
@mayank99 and I had removed this in a previous commit (https://github.com/iTwin/iTwinUI/pull/1942/commits/3861266a09f6d06d6a5f79e6e77261b56b12f309). But I brought it back since I now remember why I had it in the first place. It was because if we don't update the active index upon hover/focus, then Floating UI's `useListNavigation` keeps the stale `activeIndex`.

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/29fb5fc2-3ce0-4761-991f-6800f6edca9b

</details>

One small difference is that instead of updating the parent's activeIndex in `onFocus`, I am now doing it in `onMouseEnter`, where it is really needed because that is where we're manually changing the focus instead of letting `useListNavigation` handle it.

rohan-kadkol replied:
I realized the I was not using the `listNavigation` interaction in the `Menu` in `DropdownMenu`. Thus, the previously mentioned issue existed in the first-level `Menu` in the `DropdownMenu`.

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/bba7f760-a8ab-410e-b1e4-4589b1213fc1

</details>

This is now fixed in [`8a1bc78`](https://github.com/iTwin/iTwinUI/pull/1942/commits/8a1bc78647fe9c7d49a26f54989084279c858360). Since there are many contexts, open to suggestions (if any) for context names. In a follow up PR, I can try using [`jotai`](https://jotai.org/) to reduce the number of contexts, similar to https://github.com/iTwin/iTwinUI/pull/2011.

mayank99 replied:
Does this change mean you can remove the keyboard navigation logic from `Menu`?

mayank99 replied:
>  Since there now are many contexts

I'm not sure I understand why the contexts needed to be split? `setHasFocusedNodeInSubmenu` could be made optional instead?

rohan-kadkol replied:
> Does this change mean you can remove the keyboard navigation logic from `Menu`?

Yes, it can be removed if the children of `Menu` is always going to be a `MenuItem`. This is because `MenuItem` handles setting the parent's `listItemsRef` ([code](https://github.com/iTwin/iTwinUI/pull/1942/files#diff-4410321a0ef9f99d8d2c38b7907baf2e10f7fdf8efcff3a9ca5245db96d35db9R270-R275)) which is needed for Floating UI's`useListNavigation` to work.

But since `Menu` supports non-`MenuItem` children too, I guess some cloning will be required to add a `ref` function that updates `listItemsRef`. So, we would need to add this to all internal uses of the `Menu` component and add this cloning logic to the children inside the `Menu`.

Since some more investigation might be required, I think this can be done after this PR. If so, I can add this as an after PR TODO.

mayank99 replied:
> But since Menu supports non-MenuItem children too

This is a contrived case that will never happen.

> I think this can be done after this PR

I'm ok with this, but it should be done sooner rather than later, since we now have duplicated logic for keyboard navigation.

rohan-kadkol replied:
> This is a contrived case that will never happen.

I see, then should we update the JSDocs in a few places? Because it gives the notion that non-`MenuItem` are fully supported/allowed. Examples:

https://github.com/iTwin/iTwinUI/blob/d1af8d045f7117476abc06467d98b314b92a88c5/packages/itwinui-react/src/core/Menu/Menu.tsx#L17

https://github.com/iTwin/iTwinUI/blob/d1af8d045f7117476abc06467d98b314b92a88c5/packages/itwinui-react/src/core/ComboBox/ComboBox.tsx#L120

https://github.com/iTwin/iTwinUI/blob/d1af8d045f7117476abc06467d98b314b92a88c5/packages/itwinui-react/src/core/Select/Select.tsx#L514

> I'm ok with this, but it should be done sooner rather than later

Sounds good, added an after PR TODO with the prefix "**(Important)**" to prioritize it.

mayank99 replied:
`ComboBox` and `Select` (public) have completely different semantics from `Menu` (private), so they should not even be using `MenuItem` in the first place. It makes more sense to use [`ListItem`](https://itwinui.bentley.com/docs/list#interactive-states).

No jsdoc changes necessary in this PR, but here's what I think makes the most sense as a plan for near-future:
1. Remove redundant logic from `Menu` and abstract some parts of `DropdownMenu`/`MenuItem` into `Menu`.
2. Replace incorrect usages of `Menu` (mainly in `ComboBox` and `Select`) with a new private component (could be called `Listbox`).
3. Deprecate `MenuItem` and replace with something like `DropdownMenu.Item` (and maybe `ComboBox.Item`/`Select.Item`).
4. Update all documentation to show the correct usage of all components.

rohan-kadkol commented:
Update: I think I've addressed all review comments, all tests are now passing, and I have also updated the PR description based on the latest developments in this PR.

mayank99 commented on .changeset/fuzzy-socks-return.md:
```diff
@@ -0,0 +1,7 @@
+"@itwin/itwinui-react": minor
+---
+
+* Clicking a `MenuItem` with a non empty `submenuItems` array now toggles the submenu visibility.
+  * Thus, passing a non empty `submenuItems` array and `onClick` to `MenuItem` at the same time is not supported.
+  * If both props are passed, clicking the `MenuItem` will toggle the submenu visibility but also *still* call the `onClick` (to prevent a behavioral change).
```
"non-empty" is implied, and "supported" should be reworded as "recommended", and the warning should be mentioned.

also slight formatting nit around bullet point (since changesets will already add a bullet point).

```suggestion
Clicking a `MenuItem` with `submenuItems` now toggles the submenu visibility.
  * If both `submenuItems` and `onClick` props are passed, then clicking the `MenuItem` will toggle the submenu visibility but also *still* call the `onClick`. However, this behavior can lead to a confusing UX and is not recommended, so a warning will be shown.
```

mayank99 reviewed (approved):
LGTM.

I'm not a fan of the `useEffect` and `getFocusableNodes` in [`4f30668`](https://github.com/iTwin/iTwinUI/pull/1942/commits/4f3066820ef9ceb310eba6c5ba4a0c432260bf11), but I'm ok with it as a stopgap since all tests are now passing and our [immediate plan](https://github.com/iTwin/iTwinUI/pull/1942#discussion_r1581201279) likely involves removing all this code.
